crates/configuration/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+use anyhow::{anyhow, Result};
+use std::fs;
+
+use camino::Utf8PathBuf;
+use tempfile::{tempdir, TempDir};
+use toml::Value;
+pub const CONFIG_FILENAME: &str = "snfoundry.toml";
+
+/// Defined in snfoundry.toml
+/// Configuration not associated with any specific package
+pub trait GlobalConfig {
+    #[must_use]
+    fn tool_name() -> &'static str;
+
+    fn from_raw(config: &Value) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+/// Defined in scarb manifest
+/// Configuration associated with a specific package
+pub trait PackageConfig {}
+
+pub fn get_profile(raw_config: &Value, tool: &str, profile: &Option<String>) -> Result<Value> {
+    let profile_name = profile.as_deref().unwrap_or("default");
+    let config = raw_config
+        .get(tool)
+        .expect("Failed to find sncast config in snfoundry.toml file");
+
+    match config.get(profile_name) {
+        Some(profile_value) => Ok(profile_value.clone()),
+        None if profile_name == "default" => Ok(Value::Table(Default::default())),
+        None => Err(anyhow!("Profile [{}] not found in config", profile_name)),
+    }
+}
+
+pub fn load_global_config<T: GlobalConfig + Default>(
+    path: &Option<Utf8PathBuf>,
+    profile: &Option<String>,
+) -> Result<T> {
+    let config_path = path
+        .as_ref()
+        .and_then(|p| search_config_upwards_relative_to(p).ok())
+        .or_else(|| find_config_file().ok());
+
+    match config_path {
+        Some(path) => {
+            let raw_config = fs::read_to_string(path)
+                .expect("Failed to read snfoundry.toml config file")
+                .parse::<Value>()
+                .expect("Failed to parse snfoundry.toml config file");
+
+            let profile = get_profile(&raw_config, T::tool_name(), profile)?;
+            T::from_raw(&profile)
+        }
+        None => Ok(T::default()),
+    }
+}
+
+pub fn search_config_upwards_relative_to(current_dir: &Utf8PathBuf) -> Result<Utf8PathBuf> {
+    current_dir
+        .ancestors()
+        .find(|path| fs::metadata(path.join(CONFIG_FILENAME)).is_ok())
+        .map(|path| path.join(CONFIG_FILENAME))
+        .ok_or_else(|| {
+            anyhow!(
+                "Failed to find snfoundry.toml - not found in current nor any parent directories"
+            )
+        })
+}
+
+pub fn find_config_file() -> Result<Utf8PathBuf> {
+    search_config_upwards_relative_to(&Utf8PathBuf::try_from(
+        std::env::current_dir().expect("Failed to get current directory"),
+    )?)
+}
+
+pub trait PropertyFromCastConfig: Sized {
```

Comment:
Should this + impls names be generalized as well? I suppose these may not be cast specific, and perfectly usable for forge as well

Reply:
Done